pkg/engine/planner/physical/planner.go

```diff
@@ -1,43 +1,205 @@
 package physical
 
-import "github.com/grafana/loki/v3/pkg/engine/planner/logical"
+import (
+	"errors"
+	"fmt"
 
-// Context holds meta information needed to create a physical plan.
-type Context struct{}
+	"github.com/grafana/loki/v3/pkg/engine/planner/logical"
+)
 
 // Planner creates an executable physical plan from a logical plan.
+// Planning is done in two steps:
+//  1. Convert
+//     Instructions from the logical plan are converted into Nodes in the physical plan.
+//     Most instructions can be translated into a single Node. However, MakeTable translates into multiple DataObjScan nodes.
+//  2. Pushdown
+//     a) Push down the limit of the Limit node to the DataObjScan nodes.
+//     b) Push down the predicate from the Filter node to the DataObjScan nodes.
 type Planner struct {
-	ctx Context
+	catalog Catalog
+	plan    *Plan
```

Comment:
Out of curiosity is the only planning context the catalog and will `plan` be reused? I'm asking because all functions could be pure. Only `processMakeTable` seems to need the `Planner` struct.

Reply:
Yes, for now that's the case. The idea with `plan` was to re-use it in the second stage of planning for the filter/limit pushdown. That could possibly also be done in pure functions, though.